routing/osrm_path_segment_factory.cpp

```diff
@@ -143,40 +144,77 @@ void OsrmPathSegmentFactory(RoutingMapping & mapping, Index const & index, RawPa
     return distance(buffer.begin(), it);
   };
 
-  // Calculate estimated time for a start and a end node cases.
-  if (isStartNode && isEndNode)
+  size_t const startIndex = isStartNode ? findIntersectingSeg(startGraphNode.segment) : 0;
+  size_t const endIndex = isEndNode ? findIntersectingSeg(endGraphNode.segment) + 1 : buffer.size();
+  // @TODO(bykoianko) When start and end points belong to the same NodeID it's possible
+  // to get endIndex <= startIndex. Production version returns route building error.
+  // Debug version crashes with an assert. The test to show it USALosAnglesAriaTwentyninePalmsHighwayTimeTest.
+  LoadPathGeometry(buffer, startIndex, endIndex, index, mapping, startGraphNode, endGraphNode, isStartNode,
+                   isEndNode, loadedPathSegment);
+
+  LoadedPathSegment fullPathSegment;
+  LoadPathGeometry(buffer, 0, buffer.size(), index, mapping, FeatureGraphNode(), FeatureGraphNode(),
+                   false /* isStartNode */, false /* isEndNode */, fullPathSegment);
+  double const fullPathLengthM = PathLengthM(fullPathSegment.m_path);
+  if (fullPathLengthM == 0.0 || loadedPathSegment.m_path.empty())
+  {
+    loadedPathSegment.m_weight = 0.0;
+    return;
+  }
+
+  // Full weight (time needed to path NodeID).
+  double fullWeight = 0; /* |fullWeight * kOSRMWeightToSecondsMultiplier| is time in seconds. */
+  if (isStartNode)
+  {
+    // Note. It's possible that isStartNode == true and isEndNode == true.
+    // In that case it's ok to follow this branch of if because in that case
+    // startGraphNode.node.forward_node_id == endGraphNode.node.forward_node_id.
+    ASSERT(isStartNode != isEndNode
```

Comment:
>Why do you do assertion, if isStartNode==isEndNode is a valid case?

isStartNode==isEndNode is a valid case. 

But if start and end belongs to the save NodeID (that means isStartNode==isEndNode) 
startGraphNode.node.forward_node_id should be equal to endGraphNode.node.forward_node_id
or
startGraphNode.node.forward_node_id should be equal to endGraphNode.node.reverse_node_id

> Why do you check only forwardNodeID in the startGraphNode?

This constructor (OsrmPathSegmentFactory) is made for corner cases. There's another one (please see above) for intermediate cases. They are only called from:
```c++
  OSRMRoutingResult(Index const & index, RoutingMapping & mapping, RawRoutingResult & result)
    : m_rawResult(result), m_index(index), m_routingMapping(mapping)
  {
    for (auto const & pathSegments : m_rawResult.unpackedPathSegments)
    {
      auto numSegments = pathSegments.size();
      m_loadedSegments.resize(numSegments);
      for (size_t segmentIndex = 0; segmentIndex < numSegments; ++segmentIndex)
      {
        bool isStartNode = (segmentIndex == 0);
        bool isEndNode = (segmentIndex == numSegments - 1);
        if (isStartNode || isEndNode)
        {
          OsrmPathSegmentFactory(m_routingMapping, m_index,
                                 pathSegments[segmentIndex], m_rawResult.sourceEdge,
                                 m_rawResult.targetEdge, isStartNode, isEndNode, m_loadedSegments[segmentIndex]);
        }
        else
        {
          OsrmPathSegmentFactory(m_routingMapping, m_index,
                                 pathSegments[segmentIndex], m_loadedSegments[segmentIndex]);
        }
      }
    }
  }
```